### PR TITLE
Fix #5250 : Add support for logging device DPI events

### LINKED ISSUE #5250: [Feature Request]: Add support for logging device DPI for events
### Is your feature request related to a problem? Please describe.

Events need to include the device's (unscaled) DPI as part of its log so that we can get an idea on how devices span different density formats.

### Describe the solution you'd like

An integer for dpi (e.g. 320) should be logged as an event dimension (e.g. in EventBundleCreator). This should NOT be affected by display zoom.

### Describe alternatives you've considered

There isn't a good alternative to get display density (one would need to manually look up the device specs).

### Additional context

Needed to determine whether we need to address #360, plus it'll help with other decisions in the future.

### PR DESCRIPTION
**Pull request by Tejas-67**
## Explanation
Fix #5250 In the EventBundleCreator class, introduced a new item in the bundle to log the screen density of the device. Stored using the key "screen_density" and value retrieved from resources.displayMetrics.screenDensity . 

## Essential Checklist

- [x] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

## For UI-specific PRs only

If your PR includes UI-related changes, then:
- Add screenshots for portrait/landscape for both a tablet & phone of the before & after UI changes
- For the screenshots above, include both English and pseudo-localized (RTL) screenshots (see [RTL guide](https://github.com/oppia/oppia-android/wiki/RTL-Guidelines))
- Add a video showing the full UX flow with a screen reader enabled (see [accessibility guide](https://github.com/oppia/oppia-android/wiki/Accessibility-A11y-Guide))
- For PRs introducing new UI elements or color changes, both light and dark mode screenshots must be included
- Add a screenshot demonstrating that you ran affected Espresso tests locally & that they're passing

### CODE REVIEW

**Comment by Tejas-67** [comment]
@adhiamboperes  the tests are failing because of the increase in bundle size by 1 due to addition of screen density value.
modifying the expected bundle size in the tests would fix this. Should I go with this fix?

**Comment by Tejas-67** [comment]
@adhiamboperes please check

**Comment by adhiamboperes** [comment]
> @adhiamboperes the tests are failing because of the increase in bundle size by 1 due to addition of screen density value. modifying the expected bundle size in the tests would fix this. Should I go with this fix?


Hi @Tejas-67, I took a look -- yes, modifying the expected size in tests should be the correct way to address this.

**Comment by Tejas-67** [comment]
@adhiamboperes please check

**Comment by Tejas-67** [comment]
@adhiamboperes

File: utility/src/main/java/org/oppia/android/util/logging/EventBundleCreator.kt [hunk 1]
```diff
@@ -129,6 +129,9 @@ class EventBundleCreator @Inject constructor(
   private val appVersionCode by lazy { context.getVersionCode() }
   private val appVersionName by lazy { context.getVersionName() }
   private val eventCount by lazy { AtomicInteger() }
+  private val screenDensity by lazy {
+    context.resources.displayMetrics.densityDpi
+  }
```
Comment (adhiamboperes) [inline utility/src/main/java/org/oppia/android/util/logging/EventBundleCreator.kt]: From the docs for this API:
Screenshot 2024-01-03 at 23 33 03

It looks like this value may be affected by the user's display settings, including display zoom.  The solution required should NOT be affected by display zoom, and according to the docs, we need to use `windowManager.defaultDisplay.getMetrics(metrics)`

 If you want to obtain the physical DPI without being affected by display zoom, you need to use the DisplayMetrics method xdpi (horizontal pixels per inch) and ydpi (vertical pixels per inch):
 
```Kotlin
val metrics = DisplayMetrics()
windowManager.defaultDisplay.getMetrics(metrics)
val xDpi: Float = metrics.xdpi
val yDpi: Float = metrics.ydpi

// Convert to integer as needed
val xDpiInt: Int = xDpi.roundToInt()
val yDpiInt: Int = yDpi.roundToInt()



```

**Review by adhiamboperes — CHANGES_REQUESTED** [review]
Thanks @Tejas-67!

I have left an improvement suggestion but the overall approach is in the right direction. PTAL.

File: utility/src/main/java/org/oppia/android/util/logging/EventBundleCreator.kt (hunk 1 above)
Comment (Tejas-67) [reply utility/src/main/java/org/oppia/android/util/logging/EventBundleCreator.kt]: hey @adhiamboperes, do we need to log 2 values? Couldn't find any way to get the screen density in dpi using the xDpi and yDpi.

Comment (adhiamboperes) [reply utility/src/main/java/org/oppia/android/util/logging/EventBundleCreator.kt]: @Tejas-67, I see your concern. I have referred again to the screenshot above and it is suggested that we can get density from  `Configuration().densityDpi`, and where the information is not available(this is possible), we can set  the value to `Configuration.DENSITY_DPI_UNDEFINED`

You can update this to: 
```
 private val screenDensity by lazy {
   Configuration().densityDpi
  }
```

**Comment by adhiamboperes** [comment]
@Tejas-67, any update on this? Please assign this PR to me once it ios ready for a final review.

**Comment by Tejas-67** [comment]
hey @adhiamboperes, i have done the necessary changes. Please check.

**Review by adhiamboperes — APPROVED** [review]
Thanks @Tejas-67. This LGTM.

File: utility/src/main/java/org/oppia/android/util/logging/EventBundleCreator.kt (hunk 1 above)
Comment (adhiamboperes) [reply utility/src/main/java/org/oppia/android/util/logging/EventBundleCreator.kt]: Please check section 2 and 3 of PR guidelines for future reference: https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR

**Pull request merged**